TC plugin: add Alpaca

gzotti opened the pull request:
<!--- Provide a general summary of your changes in the Title above -->

### Description
<!--- Please include a summary of the change and which issue is fixed. -->
<!--- Please also include relevant motivation and context. -->
<!--- List any dependencies that are required for this change. -->

This branch shall be used to add support for ASCOM/Alpaca

Motivation: Our ASCOM support was reportedly prone to frequent crashes when built with Qt6. 
Meanwhile ASCOM7 is out, which seems to work again. 

Alpaca, a platform independent IP based protocol, is the new player in the ASCOM universe. We also should observe the EOL of .NET3.5 in early 2029.  So, replacing COM with that seemed the longer-time solution. 

The Alpaca standard is platform independent, so it would be a new client for all platforms. When Alpaca works and ASCOM still occasionally fails, we can then bury it. 

However, _one part of the Stellarium telescope protocol seems problematic_. The isAlive test which is called every frame exchanges a data package to the connected device. May be OK for a private serial line, but maybe not in the local IP network, connecting to semi-remote instruments controlled by miniPCs or low-power Raspberries. I cannot say at this time whether this data exchange can be moved to a lower-priority e.g. timer-triggered call. **Should therefore be developed and field tested by somebody with access and regular use of such hardware.** This change may affect all other clients. 


Fixes # (issue) [uncountable]

### Screenshots (if appropriate):

### Type of change
<!--- What types of changes does your code introduce? Put an `x` in all the boxes that apply: -->
- [x] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [x] Breaking change (fix or feature that would cause existing functionality to change)
- [x] This change requires a documentation update
- [ ] Housekeeping

### How Has This Been Tested?
<!--- Please describe in detail how you tested your changes. -->
<!--- Include details of your testing environment, and the tests you ran to -->
<!--- see how your change affects other areas of the code, etc. -->

**Test Configuration**:
* Operating system: Windows 11
* Graphics Card: Probably irrelevant

## Checklist:
<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [ ] My code follows the [code style](http://stellarium.org/doc/head/codingStyle.html) of this project.
- [ ] I have performed a self-review of my own code
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation (header file)
- [ ] I have updated the respective chapter in the Stellarium User Guide
- [ ] My changes generate no new warnings
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] New and existing unit tests pass locally with my changes
- [ ] Any dependent changes have been merged and published in downstream modules

10110111 commented:
I suggest that we keep the legacy ASCOM for Qt5 (disabled for Qt6) and just add one more protocol that would be supported regardless of toolkit.

gzotti commented:
If ASCOM7 now works well with Qt6, we can reactivate it. But somebody with actual equipment has to start testing, not reverting to 0.20 on the slightest problem. So, in principle, the first commit here can be cherry-picked into master (and 25.1) if we want that.

And yes, Alpaca also needs to come. However we can (if it makes a difference) limit that to Qt6.

10110111 commented:
> If ASCOM7 now works well with Qt6

Why would it? It's still based on COM, and this is exactly what conflicts with Qt, not the actual payload API being called.

gzotti commented:
We never actually identified what is the problem with ASCOM6.5, or did we? And now with ASCOM7 I cannot trigger a crash so far. Allegedly you just would run the simulator and some other program (SharpCap in one report, but it was not required to trigger the silent crash/vanishing), switching between them, having Stellarium in the background for a few minutes, then switching back. Unfortunately I cannot find the report that describes it best. So, for now, I'd like to have clear instructions to trigger a crash, else it could come back any time.

10110111 commented:
> We never actually identified what is the problem with ASCOM6.5, or did we?

We did, more or less, I remember having discussed this while my experimental `ascom-crash-fix` branch existed. But GitHub has an awful search engine, so I can't find anything about this now.

10110111 commented:
So a manual search over a hundred issues yielded #2874, particularly [this comment](https://github.com/Stellarium/stellarium/issues/2874#issuecomment-2161672815).

gzotti commented:
Ah, that branch, right. I just could not find it. While you did a lot of diagnosis there, there was no final test with ASCOM7. Regardless, I keep it active in this branch to allow further testing until proven faulty again (it "should not" :-) disturb normal operation without an ASCOM telescope), but will try to add Alpaca.

gzotti commented:
I am not a regular user of CdC. But for a test I installed it and activated tracking the same ASCOM simulator telescope. Now I can trigger a slew in Stellarium and observe the simulator telescope pointer moving in CdC, and vice versa. So far no issues.

gzotti commented:
After days of nocturnal hibernation and trying a few slews each evening, switching eagerly between applications, I never experienced the crash that was present with ASCOM6.5. @alex-w, @10110111 shall I reactivate ASCOM with Qt6 in master, ready for 25.1, just requiring ASCOM7? Either it works now again for all, or it still doesn't work, we can only find out when testing with equipment combinations probably none of us has or uses. 

The Alpaca client will not be ready in 25.1, but I am reading the docs to prepare.

alex-w commented:
> After days of nocturnal hibernation and trying a few slews each evening, switching eagerly between applications, I never experienced the crash that was present with ASCOM6.5. @alex-w, @10110111 shall I reactivate ASCOM with Qt6 in master, ready for 25.1, just requiring ASCOM7? Either it works now again for all, or it still doesn't work, we can only find out when testing with equipment combinations probably none of us has or uses.

Is it possible to get and check sufficient the version of ASCOM for  and show alert dialog (Stellarium required ASCOM 7+) with blocking connection to ASCOM devices when version is lesser than required?

10110111 commented:
> I never experienced the crash that was present with ASCOM6.5

Could you try, just to be sure that it's really the new version that fixed things and not your PC, whether you reproduce the crash if you replace ASCOM7 with ASCOM6.5?

> shall I reactivate ASCOM with Qt6 in master, ready for 25.1, just requiring ASCOM7?

If it's really the version that matters, and there is a way to check it, I'd make the requirement hard, e.g. saying about incompatibility in an _error_ message, refusing to make any connection until the user fixes the installation.

gzotti commented:
The latter requirement is fulfilled by modifying the test in ASCOMSupport::isASCOMSupported().

If I find time, I can install ASCOM 6.5 on yet another PC (2012 gaming notebook, W10/deprecated...) and try to trigger an error. Just that these spurious errors need to be triggered by actually operating the program in "observing simulation conditions", i.e., needs time. But maybe I have a result next weekend then.

gzotti committed:
Create Alpaca client as copy of ASCOM client.
- removed all COM stuff
- disabled any functionality
- just compiles without function.
- start point of developing further from Alpaca specs

gzotti commented:
Intermediate update. Built this with Qt6.5, ran with ASCOM 6.6, plus v23.4, both driving same ASCOM simulator and seeing each other's slew actions. After a day-long standby, 23.4 finally crashed, but seconds later also this build "stopped working". Now updated to ASCOM 7, let's see.

gzotti commented:
With ASCOM7 and running both versions plus Meshlab (just for another OpenGL app), 23.4 crashed twice, this build survived. I recommend adding ASCOM7 support to RC1 (the first 3 commits in this branch) and ask actual telescope users (bug reporters) for a test run with their real gear. If it still fails, we then know. @alex-w , @10110111 ?

alex-w commented:
> With ASCOM7 and running both versions plus Meshlab (just for another OpenGL app), 23.4 crashed twice, this build survived. I recommend adding ASCOM7 support to RC1 (the first 3 commits in this branch) and ask actual telescope users (bug reporters) for a test run with their real gear. If it still fails, we then know. @alex-w , @10110111 ?

It's OK for me

10110111 commented:
What happens if the user has ASCOM6? Do they get prompted to upgrade?

gzotti commented:
With ASCOM<7 there is no option to configure a new ASCOM telescope. However, I inadvertently had built and launched this while ASCOM6.6 was installed.  It occurred that an already configured ASCOM telescope simulator was available. After deleting it, I could not add a new one as ASCOM was not displayed as valid option. 

If you think it is really worth the effort, I could add a QMessagePanel when first activating the plugin that tests for ASCOM version number and shows a popup warning to upgrade or ignore warning and continue into the unmarked minefield. But I hope telescope users keep their gear up to date and most should run ASCOM7 by now, as it brought significant improvements for them.

10110111 commented:
> But I hope telescope users keep their gear up to date

I wouldn't rely on this. Their gear may be up to date (lubricated, aligned etc.), but the software might be what worked last time. It's better to urge the users to upgrade if our stability depends on it.

alex-w commented:
@gzotti please fix compiling in non-Windows operating systems

gzotti commented:
The plugin itself is not ready of course, only the (now) first two commits could be picked. Building the main program reportedly completed. I will see what is wrong with tests.

gzotti committed:
Fix compilation of Alpaca tests

alex-w commented:
I'm interesting how you will tests connection to Alpaca server? I fear unit tests is not applicable here...

gzotti commented:
This is highly likely. As stated, I copied the ASCOM client for a start (because the mental concept of Apaca is equal to ASCOM) and just replaced the letters "ASCOM" by "Alpaca", then disabled all COM stuff. This plugin extension IS IN NO WAY ready for review or automatic tests. What I propose is REACTIVATING ASCOM support as ASCOM7 SEEMS TO BE stable. This can be done by cherry-picking the first two commits of this unfinished branch to master. Nothing that says Alpaca currently is expected to work before I switch to "Ready for merge".

10110111 commented:
> What I propose is REACTIVATING ASCOM support

After the recently added warning message box I'm OK with this.

gzotti committed:
Suppress probably useless test to make CI happy for now

gzotti committed:
cleanup

artooksanen commented:
Here is a minimal python test client I have been using for a Alpaca/Alpyca telescope driver:

```
from alpaca.telescope import *

T = Telescope('192.168.8.102:5555', 0)
T.Connect()
print(f'Connected to {T.Name}',T.Connected)
ra=T.RightAscension
print("ra=",ra)  
de=T.Declination
print("de=",de)  
T.Disconnect()
```

gzotti commented:
Thanks. Right, studying the Python package and client may provide some insights what's actually necessary. Then the first great hurdle is the discovery protocol.

10110111 commented:
There's already been some python client, someone opened a Discussion or something a month or two ago.

10110111 commented:
> someone opened a Discussion or something

Here it is: https://github.com/Stellarium/stellarium/discussions/3823#discussioncomment-10520922.

gzotti committed:
BAD COMMIT
This breaks everything ASCOM attempting to upgrade to ASCOM7 protocol
